ra/ra.go

```diff
@@ -1376,24 +1376,36 @@ func (ra *RegistrationAuthorityImpl) checkCertificatesPerFQDNSetLimit(ctx contex
 		return nil
 	}
 
-	issuanceTimestampsInWindow, err := ra.SA.FQDNSetTimestampsForWindow(ctx, &sapb.CountFQDNSetsRequest{
+	issuance, err := ra.SA.FQDNSetTimestampsForWindow(ctx, &sapb.CountFQDNSetsRequest{
 		Domains: names,
 		Window:  limit.Window.Duration.Nanoseconds(),
 	})
 	if err != nil {
 		return fmt.Errorf("checking duplicate certificate limit for %q: %s", names, err)
 	}
 
-	if int64(len(issuanceTimestampsInWindow.Timestamps)) >= threshold {
-		// FQDNSetTimestampsForWindow returns the issuance timestamps in
-		// ascending order so we can assume that the first one is the oldest.
-		retryAfter := time.Unix(0, issuanceTimestampsInWindow.Timestamps[0]).Add(limit.Window.Duration)
+	if int64(len(issuance.Timestamps)) < threshold {
+		// Issuance in window is below the threshold, no need to limit.
+		return nil
+	} else {
+		// Evaluate the rate limit using a leaky bucket algorithm. The bucket
+		// has a capacity of threshold and is refilled at a rate of 1 token per
+		// limit.Window/threshold from the time of each issuance timestamp.
+		nsPerToken := limit.Window.Nanoseconds() / threshold
+		for i, timestamp := range issuance.Timestamps {
+			nsUntilNextToken := int64(i+1) * nsPerToken
+			nsSinceIssuance := ra.clk.Now().Sub(time.Unix(0, timestamp)).Nanoseconds()
+			if nsSinceIssuance > nsUntilNextToken {
+				// Found a token.
```

Comment:
Nit: This should be something like: "Found an interval during which more tokens were 'generated' than consumed."